pkg/endpoint/endpoint.go

```diff
@@ -2422,8 +2426,24 @@ func (e *Endpoint) syncPolicyMap() error {
 
 	currentMapContents, err := e.PolicyMap.DumpToSlice()
 
+	// If map is unable to be dumped, attempt to close map and open it again.
 	if err != nil {
-		return fmt.Errorf("unable to dump PolicyMap for endpoint: %s", err)
+		// Close to avoid file descriptor leaking.
+		err := e.PolicyMap.Close()
+		if err != nil {
+			return fmt.Errorf("unable to close PolicyMap which was not able to be dumped: %s", err)
```

Comment:
In the cases where we see an error from `DumpToSlice` isn't it also likely that we would see an error with `Close`? Would we ever get past this return to actually try to reopen the map?

Reply:
This is something I was wondering myself. The problem I am having with trying to 'fix' this PR is that I have not been able to reproduce the issue locally, and need more data to fully determine the cause of the problem. Given that we can't access the map anyway in the cases where we get this 'bad file descriptor' error, I figure that if this does function in the way that you suspect, that we aren't in any worse shape than we are now, and can change the code accordingly. I'm hoping that additional logging can help provide insight into this as well :/

Reply:
I understand, the goal is to improve the probability of recovery and reliability. Not doing anything is the worst decision we can make here.

@raybejjani is correct that the Close() will likely fail as well. We should log an error that includes the value of the fd and then continue to re-open. If the map is really still open, close will not fail unless the fd is invalid. If the fd is invalid but not released there is nothing we can do except exiting the agent eventually. The best way to see if need to do that eventually is to log an error but continue.

Reply:
Fixed in latest upload. `PolicyMap.Close()` will log the file descriptor for the map being closed, and now an error is logged instead of returning one.